postproc/postproc/trace.py

```diff
@@ -0,0 +1,295 @@
+                    f"The Trace {i.name} is not present in the DataFrame keys {df.keys()}. Skipped."
+                )
+                pass
+
+    def derive_raw_data(self, traces):
+        """Derive the current self.raw_traces from other traces
+
+        Using the other traces provided we evaluate the expression provided in derivation_params.
+        derivation_params has the following structure:
+
+        [operation, [trace_names], other_arguments]
+
+        and the derivation is computed so that:
+
+        operation(*[i for i in traces[trace_names]], *other_arguments)
+
+        Args:
+            - traces (dict): Complete dictionary of traces from TraceDB
+        """
+        if not self.derivation_params:
+            return
+
+        [op, root_traces, *args] = self.derivation_params
+
+        for file in traces[root_traces[0]].raw_traces:
+            input_traces = [traces[i].raw_traces[file] for i in root_traces]
+
+            if hasattr(Utils, op) and callable(func := getattr(Utils, op)):
+                derived_trace = func(*input_traces, *args)
+            elif hasattr(numpy, op) and callable(func := getattr(numpy, op)):
+                derived_trace = func(*input_traces, *args)
+            elif hasattr(scipy, op) and callable(func := getattr(scipy, op)):
+                derived_trace = func(*input_traces, *args)
+            else:
+                raise TraceError(
+                    f"The derived trace {self.name} presents an unknown operation {op}."
+                )
+
+            self.raw_traces[file] = self.multi * numpy.array(derived_trace)
+
+    def refine(self, time_trace={}):
+        """Fill refined_traces using evaluating the keys as member functions in Utils, numpy or scipy
+
+        The refine_traces key can be a str or a list of arguments. In case of a str, it is evaluated as a member
+        function of Utils, numpy or scipy. In case of a list, the first term is the member function while the others
+        are arguments to be fed after the traces.
+        """
+
+        logging.info(f"Refining {self.name}")
+
+        for op_key in self.refined_traces.columns:
+            tr = []
+            if len(self.refined_traces[op_key]) == 1 and isinstance(
+                self.refined_traces[op_key][0], str
+            ):
+                f = open(self.refined_traces[op_key][0])
+                for l in f.readlines():
+                    try:
+                        tr.append(self.multi * float(l))
+                    except:
+                        pass
+
+            else:
+
+                for file, trace in self.raw_traces.items():
+
+                    logging.info(f"Processing {file}")
+
+                    op_args = []
+                    op = op_key
+                    try:
+                        l = eval(op_key)
+                        op = l[0]
+                        op_args = l[1:]
+                    except NameError:
+                        pass
```

Comment:
ok, it is totally doable, I verified. Tuples work as keys for pandas df and dicts. It requires some rework in the code because I was using the fact that this is a string to easily print picture titles. I will try

Reply:
after some implementing I have found where is the real problem: parquet requires string-only column names. I will revert to old version